test/performance/latency_test.go

```diff
@@ -77,50 +81,41 @@ func timeToServe(t *testing.T, img, query string, reqTimeout time.Duration) {
 		t.Fatalf("Error probing domain %s: %v", domain, err)
 	}
 
-	endpoint, err := spoof.ResolveEndpoint(clients.KubeClient.Kube, domain, test.ServingFlags.ResolvableDomain,
+	url, err := spoof.ResolveEndpoint(clients.KubeClient.Kube, domain, test.ServingFlags.ResolvableDomain,
 		pkgTest.Flags.IngressEndpoint)
 	if err != nil {
 		t.Fatalf("Cannot resolve service endpoint: %v", err)
 	}
 
-	opts := loadgenerator.GeneratorOptions{
-		Duration:       duration,
-		NumThreads:     1,
-		NumConnections: 5,
-		Domain:         domain,
-		URL:            fmt.Sprintf("%s/?%s", endpoint, query),
-		RequestTimeout: reqTimeout,
-		LoadFactors:    []float64{1},
-		FileNamePrefix: tName,
-	}
-	var flags []int
-	if !test.ServingFlags.ResolvableDomain {
-		flags = append(flags, loadgenerator.AddHostHeader)
-	}
-
-	resp, err := opts.RunLoadTest(flags...)
-	if err != nil {
-		t.Fatalf("Generating traffic via fortio failed: %v", err)
+	if !strings.HasPrefix(url, httpPrefix) {
+		url = httpPrefix + url
 	}
 
-	// Save the json result for benchmarking
-	resp.SaveJSON()
-
-	if len(resp.Result) == 0 {
-		t.Fatal("No result found for the load test")
+	headers := make(map[string][]string)
+	if !test.ServingFlags.ResolvableDomain {
+		headers["Host"] = []string{domain}
 	}
 
-	if resp.ErrorsPercentage(0) > 0 {
-		t.Fatal("Found non 200 response")
+	pacer := vegeta.ConstantPacer{Freq: baseQPS, Per: time.Second}
+	targeter := vegeta.NewStaticTargeter(vegeta.Target{
+		Method: http.MethodGet,
+		Header: headers,
+		URL:    url + "?" + query,
+	})
+	attacker := vegeta.NewAttacker()
+
+	var metrics vegeta.Metrics
+	for res := range attacker.Attack(targeter, pacer, duration, tName) {
+		metrics.Add(res)
 	}
+	metrics.Close()
 
 	// Add latency metrics
 	var tc []junit.TestCase
-	for _, p := range resp.Result[0].DurationHistogram.Percentiles {
-		val := float32(p.Value) * 1000
-		name := fmt.Sprintf("p%d(ms)", int(p.Percentile))
-		tc = append(tc, perf.CreatePerfTestCase(val, name, tName))
-	}
+	// Add latency metrics
+	tc = append(tc, perf.CreatePerfTestCase(float32(metrics.Latencies.P50.Seconds()*1000), "p50(ms)", tName))
+	tc = append(tc, perf.CreatePerfTestCase(float32(metrics.Latencies.Quantile(0.90).Seconds()*1000), "p90(ms)", tName))
+	tc = append(tc, perf.CreatePerfTestCase(float32(metrics.Latencies.P99.Seconds()*1000), "p99(ms)", tName))
```

Comment:
I know it is unrelated but it would be good to also collect the 99.9 percentile throughout the performance tests. The chance that a user hits the 99.9 percentile is higher than it seems.

Reply:
We need to stop aggregating client-side.  Mako's facilities for all of this are far better than what we will ever build ourselves :)

Reply:
Yes, this PR is mainly focusing on replacing the load generator. There will be followup PRs to provide better visualization and analysis for the test results.